building

Hello

Thanks for making this code available. I'm probably missing something obvious here, but I'm having no luck with building the code. I'm using ubuntu 16.04

I'm in the petmr-RESOLUTE directory and have done
```
cmake . 
make 

```
Two issues so far
a) I get an error

```
In file included from /usr/include/c++/5/initializer_list:36:0,
                 from /usr/local/include/nlohmann/json.hpp:41,
                 from /mnt/edrive/MJF/resolute/petmr-RESOLUTE/src/Resolute.cpp:29:
/usr/include/c++/5/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
 #error This file requires compiler and library support \
```

b) How do I tell it where to look for ITK etc - I'm getting 
`/petmr-RESOLUTE/src/Resolute.cpp:31:32: fatal error: itkCastImageFilter.h: No such file or directory
`

Thanks for any advice

Michael

Hi Ben

I think I got the c++ error by running cmake in the src directory, so probably that isn't actually a problem.

However, I'm still having no luck with the ITK. I got that along with ANTS, and am running

`cmake . -DANTs_SOURCE_DIR=/mnt/edrive/MJF/ants/ANTs -DANTs_LIBRARY_DIR=/mnt/edrive/MJF/ants/antsbin/lib/  -DITK_DIR=/mnt/edrive/MJF/ants/antsbin/ITKv4-build`

followed by make, and I get: 
```

[ 25%] Building CXX object src/CMakeFiles/resolute.dir/Resolute.cpp.o
[ 50%] Building CXX object src/CMakeFiles/resolute.dir/mnt/edrive/MJF/ants/ANTs/Examples/antsRegistration.cxx.o
[ 75%] Building CXX object src/CMakeFiles/resolute.dir/mnt/edrive/MJF/ants/ANTs/Examples/antsApplyTransforms.cxx.o
[100%] Linking CXX executable resolute
/mnt/edrive/MJF/ants/antsbin/lib/libantsUtilities.a(antsRegistration2DDouble.cxx.o): In function `itk::GradientDescentOptimizerBasev4Template<double>::~GradientDescentOptimizerBasev4Template()':
antsRegistration2DDouble.cxx:(.text._ZN3itk38GradientDescentOptimizerBasev4TemplateIdED2Ev[_ZN3itk38GradientDescentOptimizerBasev4TemplateIdED5Ev]+0x178): undefined reference to `itk::ObjectToObjectOptimizerBaseTemplate<double>::~ObjectToObjectOptimizerBaseTemplate()'
...
```

with a whole load of undefined ITK references. Any suggestions appreciated

Thanks

Michael

Hi 

ANTs seems to build successfully. `make test `passes most tests:
```


99% tests passed, 3 tests failed out of 221

Total Test time (real) = 332.48 sec

The following tests FAILED:
	  3 - antsRegistrationTest_AffineRotationMasks (Failed)
	 15 - antsRegistrationTest_SimilarityRotationMasks (Failed)
	 25 - antsRegistrationTest_initializePerStage_ComparisonTesting (Failed)
Errors while running CTest
Makefile:149: recipe for target 'test' failed
make: *** [test] Error 8

```

Thanks

Michael

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,7 +4,11 @@ sudo: true
 
 cache:
 - directories:
-  - $HOME/antsbin
+  - $HOME/glog-build
+  - $HOME/googletest-build
+  - $HOME/ITK-BUILD
+  - $HOME/ITK-INSTALL
+  - $HOME/ANTS-BUILD
 
 addons:
   apt:
@@ -20,6 +24,9 @@ addons:
       - libgtest-dev
       - libdcmtk2-dev
       - ccache
+      - libgflags-dev
+      - libgoogle-glog-dev
+      - wget
 
 matrix:
   include:
@@ -32,36 +39,42 @@ matrix:
     env: COMPILER_NAME=gcc CXX=g++-5 CC=gcc-5
 
 before_script:
-  - g++ --version
   - if [ "$TRAVIS_OS_NAME" == "osx" ];  then echo "need cmake 3.2"; fi
-  - if [ "$TRAVIS_OS_NAME" == "osx" ]; then brew update && brew install ccache; fi
+  - if [ "$TRAVIS_OS_NAME" == "osx" ]; then brew update && brew install ccache glog ; fi
   - if [ "$TRAVIS_OS_NAME" == "osx" ]; then brew install dcmtk; fi
   - if [ "$TRAVIS_OS_NAME" == "osx" ]; then export PATH="/usr/local/opt/ccache/libexec:$PATH"; fi
-  - git clone https://github.com/google/googletest
-  - cd googletest
-  - cmake -DCMAKE_BUILD_TYPE=Release .
+  - ls ~
+  - git clone https://github.com/google/glog ~/glog
+  - if [ ! -e ~/glog-build/install_manifest.txt ]; then
+        mkdir -p ~/glog-build &&
+        cd ~/glog-build &&
+        cmake -DCMAKE_BUILD_TYPE=Release ~/glog &&
+        make -j2;
+      fi
+  - cd ~/glog-build
+  - sudo make install #step 1 to here
+  - git clone --branch=release-1.8.0 https://github.com/google/googletest ~/googletest
+  - if [ ! -e ~/googletest-build/install_manifest.txt ]; then
+        mkdir -p ~/googletest-build &&
+        cd ~/googletest-build &&
+        cmake -DCMAKE_BUILD_TYPE=Release ~/googletest &&
+        make -j2;
+    fi
+  - cd ~/googletest-build
+  - sudo make install #step 1 to here
+  - git clone --branch=v3.1.1 https://github.com/nlohmann/json.git ~/json
+  - cd ~/json
+  - cmake  -DBUILD_TESTING=OFF .
   - make -j2
-  - sudo make install
-  - cd ..
-  - git clone --branch=v0.3.5 https://github.com/google/glog
-  - cd glog
-  - cmake  -DCMAKE_BUILD_TYPE=Release .
-  - make -j2
-  - sudo make install
-  - cd ..
-  - git clone --branch=v3.1.1 https://github.com/nlohmann/json.git
-  - cd json
-  - cmake  -DCMAKE_BUILD_TYPE=Release .
-  - make -j2
-  - sudo make install
-  - cd ..
-  - if [ ! -e ~/ANTs ]; then git clone --branch=v2.2.0 https://github.com/ANTsX/ANTs.git ~/ANTs; fi
+  - sudo make install #step 2 to here
+  - cd ~
+  - git clone --branch=v2.2.0 https://github.com/ANTsX/ANTs.git ~/ANTs
+  - if [ ! -e ~/ANTS-BUILD/ANTS-build ]; then
+        mkdir -p ~/ANTS-BUILD &&
+        cd ~/ANTS-BUILD &&
+        cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING:BOOL=OFF ~/ANTs && make -j2 --keep-going;
+    fi #step 4 to here
 
 script:
-  - |
-    if [ ! -e ~/antsbin/ANTs-build ]; then
-        if [ ! -e ~/antsbin ]; then mkdir ~/antsbin; fi
-        cd ~/antsbin;
-        cmake -DCMAKE_BUILD_TYPE=Release -DRUN_LONG_TESTS=OFF -DRUN_SHORT_TESTS=OFF ~/ANTs && make -j2 --keep-going
-    fi
-  - cd ~/build/UCL/petmr-RESOLUTE && cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON -DITK_DIR=~/antsbin/ITKv4-build -DANTs_SOURCE_DIR=~/ANTs -DANTs_LIBRARY_DIR=~/antsbin/lib && make -j2 && ctest --verbose
+  - ls ~/ANTS-BUILD
+  - cd ~/build/UCL/petmr-RESOLUTE && cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON -DITK_DIR=$HOME/ANTS-BUILD/ITKv4-build -DANTs_SOURCE_DIR=$HOME/ANTs -DANTs_LIBRARY_DIR=$HOME/ANTS-BUILD/lib && make -j2 && ctest --verbose

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -3,9 +3,9 @@ add_executable(resolute Resolute.cpp
   ${ANTs_SOURCE_DIR}/Examples/antsApplyTransforms.cxx
   )
 target_link_libraries(resolute 
+      ${ANTS_LIBS}
       ${ITK_LIBRARIES}  
       ${Boost_LIBRARIES}
-      ${ANTS_LIBS}
       glog 
       nlohmann_json
     )

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -1,8 +1,8 @@
 add_executable(test_reg test_registration.cpp ${ANTs_SOURCE_DIR}/Examples/antsRegistration.cxx )
 target_link_libraries(test_reg 
+      ${ANTS_LIBS}
       ${ITK_LIBRARIES}  
       ${Boost_LIBRARIES}
-      ${ANTS_LIBS}
       glog 
       nlohmann_json
     )
@@ -21,9 +21,9 @@ file(GLOB SRCS
 add_executable(testRESOLUTE ${SRCS})
 
 target_link_libraries(testRESOLUTE
+  ${ANTS_LIBS}
   ${ITK_LIBRARIES}  
   ${Boost_LIBRARIES}
-  ${ANTS_LIBS}
   gtest
   gmock
   glog 

